src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -56,16 +100,114 @@ public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(Depe
+            }
+
+            IList<LockFileTarget> targetsList = null;
+            if (IsInstalledAndTransitiveComputationNeeded)
+            {
+                // clear the transitive packages cache, since we don't know when a dependency has been removed
+                InstalledPackages.Clear();
+                TransitivePackages.Clear();
+                targetsList = await GetTargetsListAsync(assetsPath, token);
+            }
+
+            var frameworkSorter = new NuGetFrameworkSorter();
+
+            // get installed packages
+            List<PackageReference> installedPackages = packageSpec
+                .TargetFrameworks
+                .SelectMany(f => FetchInstalledPackagesList(f.Dependencies, f.FrameworkName, InstalledPackages, targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
+
+            // get transitive packages
+            IEnumerable<PackageReference> transitivePackages = packageSpec
+                .TargetFrameworks
+                .SelectMany(f => FetchTransitivePackagesList(f.FrameworkName, InstalledPackages, TransitivePackages, targetsList))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First());
+
+            IEnumerable<TransitivePackageReference> transitivePackagesWithOrigins;
+            if (await IsTransitiveOriginExpEnabled.GetValueAsync())
+            {
+                // Get Transitive Origins
+                transitivePackagesWithOrigins = transitivePackages
+                    .Select(packageRef =>
+                    {
+                        (PackageReference pr, TransitiveEntry transitiveEntry) tupl = (packageRef, GetTransitivePackageOrigin(packageRef, installedPackages, targetsList, token));
+                        return tupl;
+                    })
+                    .Select(tuple => MergeTransitiveOrigin(tuple.pr, tuple.transitiveEntry));
+            }
+            else
+            {
+                // Get Transitive packages without Transitive Origins
+                transitivePackagesWithOrigins = transitivePackages
+                    .Select(packageRef => new TransitivePackageReference(packageRef));
+            }
+
+            List<TransitivePackageReference> transitivePkgsResult = transitivePackagesWithOrigins.ToList(); // Materialize results before setting IsInstalledAndTransitiveComputationNeeded flag to false
+            IsInstalledAndTransitiveComputationNeeded = false;
+
+            return new ProjectPackages(installedPackages, transitivePkgsResult);
+        }
+
+        protected abstract IEnumerable<PackageReference> FetchInstalledPackagesList(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, T installedPackages, IList<LockFileTarget> targets);
```

Comment:
Given this is only called once, and `InstalledPackages` is an instance property, I don't think that `T installedPackages` is required. The implementation can use the instance property directly.

Reply:
Removed params.